### PR TITLE
Implement a Time-To-Live (TTL) Indexing feature a la MongoDB

### LINKED ISSUE #346: Support MongoDB TTL feature
It would be awesome if NeDB supported a Time-To-Live (TTL) feature for expiring documents per datastore.

MongoDB has implemented this feature via a special indexing property:
    https://docs.mongodb.org/manual/tutorial/expire-data/

### PR DESCRIPTION
**Pull request by JamesMGreene**
Fixes #346

_FIRST DRAFT!_

**TODO:**
 - [ ] Filter or remove expired documents when retrieved via any index, not just via the index with the expiration built-in
 - [ ] If only filtering expired documents (above), filter or remove expired documents when compacting

### CODE REVIEW

**Comment by JamesMGreene** [comment]
P.S. Expired documents are naturally removed from the persisted file during compaction.

File: lib/indexes.js [hunk 1]
```diff
@@ -230,28 +262,21 @@ Index.prototype.revertUpdate = function (oldDoc, newDoc) {
...
 
-// Append all elements in toAppend to array
-function append (array, toAppend) {
-  var i;
-
-  for (i = 0; i < toAppend.length; i += 1) {
-    array.push(toAppend[i]);
-  }
-}
-
```
Comment (JamesMGreene) [inline lib/indexes.js]: The above was a random chunk of dead code.

**Comment by JamesMGreene** [comment]
> P.S. Expired documents are naturally removed from the persisted file during compaction.

Whoops, looks like I was incorrect on that one.  I forgot that the expiration only affects indices that have it enabled, so it won't affect the `_id` index that is used to get all documents for persistence.

@louischatriot Any tips on how to deal with removing documents that are expired?  I was having trouble trying to find a natural place to put that without violating the principle of separation of concerns too much &mdash; as it has to be in either "datastore.js" or "persistence.js" rather than "indexes.js".

**Comment by lukasbestle** [comment]
As far as I can tell, it's not possible to define a TTL for single documents either. Rather than implementing it with indices, wouldn't it be more robust to make this a new "special" field that is prefixed with a dollar symbol and specific to each document?

**Comment by JamesMGreene** [comment]
@lukasbestle: Sounds like a lot of storage overhead but it is probably easier to implement and perhaps more robust.

**Comment by JamesMGreene** [comment]
FYI, I have a new implementation for this following @lukasbestle's suggestion that seems to be working out much better. Just adding some tests, should be able to submit it later tonight as a separate PR.

**Pull request closed without merging**